### DISCUSSION TITLE
Both Scrolling and Fixed Text

### DISCUSSION THREAD

**Discussion by agganitk** [board707/DMD_STM32#58] [General]
I want to have some text on left most side FIXED and some text keep on scrolling on remaning screen towards right.
So my display will have both Fixed and Scrolling text at same time.
But when I run dmd.stepMarquee function whole screen starts scrolling-the text I want fixed also starts scrolling with text I want to scroll.
please suggest.

**Comment by board707** [comment]
Hi
You need to use dual buffer feature. 
At every step of the scrolling - first move the scrolling text and next print the fixed one above the scrolling.

**Reply by agganitk** [reply]
Can you please share some example of dual buffer feature.  I am not sure how to use it. 
i did like below.. good news is character 'A' stays but some LEDs just next to A character are flickering.

while(!ret){
      if ((millis() - prev_step) > 150 ) {
       ret=dmd.stepMarquee(1,0);

       dmd.drawChar(1, 1, 'A', GRAPHICS_NORMAL);
       prev_step=millis();
       //delay(1000);
     }

**Reply by board707** [reply]
Example sketch for fixed and scrolling text:
```
/*--------------------------------------------------------------------------------------
  Draw fixed and scrolling text in the same screen

  DMD_STM32a example code for STM32F103xxx board
  ------------------------------------------------------------------------------------- */
#include "DMD_RGB.h"

// Fonts includes
#include "st_fonts/Arial14.h"

//Number of panels in x and y axis
#define DISPLAYS_ACROSS 1
#define DISPLAYS_DOWN 1

// Enable of output buffering
// if true, changes only outputs to matrix after
// swapBuffers(true) command
// If dual buffer not enabled, all output draw at matrix directly
// and swapBuffers(true) cimmand do nothing
#define ENABLE_DUAL_BUFFER true


// ==== DMD_RGB pins ====
// mux pins - A, B, C... all mux pins must be selected from same port!
#define DMD_PIN_A PB6
#define DMD_PIN_B PB5
#define DMD_PIN_C PB4
#define DMD_PIN_D PB3
#define DMD_PIN_E PB8
// put all mux pins at list
uint8_t mux_list[] = { DMD_PIN_A , DMD_PIN_B , DMD_PIN_C , DMD_PIN_D , DMD_PIN_E };

// pin OE must be one of PB0 PB1 PA6 PA7
#define DMD_PIN_nOE PB0
#define DMD_PIN_SCLK PB7

// Pins for R0, G0, B0, R1, G1, B1 channels and for clock.
// By default the library uses RGB color order.
// If you need to change this - reorder the R0, G0, B0, R1, G1, B1 pins.
// All this pins also must be selected from same port!
uint8_t custom_rgbpins[] = { PA6, PA0, PA1, PA2, PA3, PA4, PA5 }; // CLK, R0, G0, B0, R1, G1, B1

// Fire up the DMD object as dmd<MATRIX_TYPE, COLOR_DEPTH>
// We use 64x32 matrix with 16 scans and 4bit color:
DMD_RGB <RGB64x32plainS16, COLOR_4BITS_Packed> dmd(mux_list, DMD_PIN_nOE, DMD_PIN_SCLK, custom_rgbpins, DISPLAYS_ACROSS, DISPLAYS_DOWN, ENABLE_DUAL_BUFFER);
// other options are:
// <RGB32x16plainS8> -  32x16 matrix with 8scans
// <RGB80x40plainS20> - 80x40 matrix with 20scans
// <RGB64x64plainS32> - 64x64 matrix with 32scans
// Color depth - <COLOR_4BITS_Packed>, <COLOR_4BITS> or <COLOR_1BITS>


// --- Define fonts ----
// DMD.h old style font
DMD_Standard_Font Arial_F(Arial_14);

void setup(void)
{
  dmd.init();
  dmd.setBrightness(200);
  dmd.disableFastTextShift(true);

}

void loop(void)
{
  // create foreground colors
  uint16_t col[] = {
    dmd.Color888(255, 0, 0), // red
    dmd.Color888(0, 255, 0), // green
    dmd.Color888(0, 0, 255)  // blue

  };
  uint16_t bg = 0;  // background - black
  int col_cnt = 3;   // color count

  // text
  char m[30] = "Hello Arduino!";
  dmd.selectFont(&Arial_F);
  dmd.setTextColor(col[0], bg);

  // shift steps in pixels for running text (positive - shift right, negative - left)
  int8_t step[] = { 1, -1, -2, 2 };
  // running text shift interval
  uint16_t interval = 30;

  long prev_step = millis();
  uint8_t col_ptr = 0;
  uint8_t i = 0, b = 0;
  uint8_t test = 255;
  uint8_t test_cnt = 4;


  dmd.drawMarqueeX(m, -1 * (dmd.stringWidth(m)), 0);
  dmd.drawStringX(0, 0, "23", col[1]);
  dmd.swapBuffers(true);


  while (1) {
    if ((millis() - prev_step) > interval) {

      // scrolling text - update it BEFORE fixed text
      if (dmd.stepMarquee(step[i], 0) & 1) { // if text is reached screen bounds

        // restart marquee after reaching the screen bounds
        dmd.clearScreen(true);
        // set next text color
        col_ptr++;
        if (col_ptr >= col_cnt) col_ptr = 0;
        dmd.setTextColor(col[col_ptr], bg);
        // set new text speed
        i++;
        if (i > 3) i = 0;
        // draw new marquee
        if (step[i] < 0) dmd.drawMarqueeX(m, dmd.width() - 1, 0);
        else dmd.drawMarqueeX(m, -1 * dmd.stringWidth(m), 0);
      }

      // draw fixed text AFTER scroll update
      dmd.drawFilledBox(0, 0, 15, dmd.height() - 1, bg);
      dmd.drawStringX(0, 0, "23", col[1]);

      // output mem buffer to matrix
      dmd.swapBuffers(true);
      prev_step = millis();

    }
  }
}
```

**Reply by board707** [reply]
And video
[https://youtu.be/DoOcfGb0PLw](https://youtu.be/DoOcfGb0PLw)

**Comment by board707** [comment]
> I see you have mentioned Roger Clarks's library need to download.

It is not the library, it is STM32 Arduino support package. 
You should remove the STM32 support, installed in your IDE and follow this instructions:
https://github.com/rogerclarkmelbourne/Arduino_STM32/wiki/Installation

**Reply by agganitk** [reply]
yess  thanks  i relaized this..   now doing this..

**Reply by agganitk** [reply]
I did but when I compile dmd_monochrome.. i see belwo errors.
If I don't add Adafruit_GFX_Library then i see 
:\Users\INTEL\Documents\Arduino\libraries\DMD_STM32-dev-V2/DMD_STM32a.h:45:26: fatal error: Adafruit_GFX.h: No such file or directory

 #include "Adafruit_GFX.h"

                          ^

but if i add Adafruit_GFX_Library folder in library then other errors such as
 #include <Adafruit_I2CDevice.h>

**Reply by board707** [reply]
Please use Adafruit_GFX libary version 1.7
https://github.com/adafruit/Adafruit-GFX-Library/releases/tag/1.7.0

**Comment by agganitk** [comment]
I done everything but yet display is not working. 
I checked all pins using logic analyzer. I see PA15 pin is not toggling and is always High.  
#define DMD_PIN_SCLK PA15       

Rest all pins (A, B , OE and data) are toggling as expected. 

I run dmd_monocrhome program with commenting the  #define DMD_PARA   line.

**Reply by board707** [reply]
Please show all your code above setup function.

**Reply by agganitk** [reply]
/*--------------------------------------------------------------------------------------
 Demo code for DMD panels

 DMD_STM32a example code for STM32F103xxx board
 ------------------------------------------------------------------------------------- */
// choose between Parallel ans SPI wiring
// comment line below for SPI connection
//#define DMD_PARA
/*--------------------------------------------------------------------------------------
  Includes
--------------------------------------------------------------------------------------*/
#if defined(DMD_PARA)
#include "DMD_Monochrome_Parallel.h"
#else
#include "DMD_MonoChrome_SPI.h"
#endif

// Fonts includes
#include "st_fonts/UkrRusArial14.h"
#pragma GCC diagnostic ignored "-Wnarrowing"
#pragma GCC diagnostic ignored "-Woverflow"
#include "gfx_fonts/GlametrixLight12pt7b.h"
#include "gfx_fonts/GlametrixBold12pt7b.h"
#pragma GCC diagnostic warning "-Wnarrowing"
#pragma GCC diagnostic warning "-Woverflow" 

//Number of panels in x and y axis
#define DISPLAYS_ACROSS 1
#define DISPLAYS_DOWN 2

// Enable of output buffering
// if true, changes only outputs to matrix after
// swapBuffers(true) command
// If dual buffer not enabled, all output draw at matrix directly
// and swapBuffers(true) cimmand do nothing
#define ENABLE_DUAL_BUFFER true

// ----- Select pins for P10 matrix connection ------------
// pins A, B, SCLK may be any digital I/O, pin nOE should be Timer3 PWM pin as PB0,PB1
// if connect as SPI, do not use corresponding MiSO pin - PA6 for SPI1 and PB14 for SPI2
#define DMD_PIN_A PA1
#define DMD_PIN_B PA4
#define DMD_PIN_nOE PB1
//#define DMD_PIN_nOE PA6
#define DMD_PIN_SCLK PA15
//#define DMD_PIN_SCLK PB4


// pins for SPI connect
// SPI specific pins as CLK and R_DATA has predefined values:
//   for SPI(1) CLK = PA5  R_DATA = PA7
//   for SPI(2) CLK = PB13  R_DATA = PB15
// --------------------------------------------------------

//=== Config for Parallel connect ====
#if defined(DMD_PARA)
//pins for rows at x axis
// example for two rows
// all those pins must be selected from same port!
uint8_t pins[] = { PA5, PA7, PA6 };  // CLK , row1, row 2

//Fire up the DMD object as dmd
DMD_Monochrome_Parallel dmd(DMD_PIN_A, DMD_PIN_B, DMD_PIN_nOE, DMD_PIN_SCLK, pins, DISPLAYS_ACROSS, DISPLAYS_DOWN,ENABLE_DUAL_BUFFER);
#else

//=== Config for SPI connect ====
SPIClass dmd_spi(1);
DMD_MonoChrome_SPI dmd(DMD_PIN_A, DMD_PIN_B, DMD_PIN_nOE, DMD_PIN_SCLK, DISPLAYS_ACROSS, DISPLAYS_DOWN, dmd_spi, ENABLE_DUAL_BUFFER);
#endif

// --- Define fonts ----
// DMD.h old style font
DMD_Standard_Font UkrRusArial_F(UkrRusArial_14);

// GFX font with sepatate parts for Latin and Cyrillic chars
DMD_GFX_Font GlametrixL((uint8_t*)&GlametrixLight12pt7b,(uint8_t*)&GlametrixLight12pt8b_rus,0x80,13);


/*--------------------------------------------------------------------------------------
  UTF8 char recoding
 
--------------------------------------------------------------------------------------*/
int utf8_rus(char* dest, const unsigned char* src) {
  
  uint16_t i, j;
  for ( i =0, j =0; src[i]; i++) {
   if ((src[i] == 0xD0 )&& src[i+1])  { dest[j++] = src[++i] - 0x10;}
    else if ((src[i] == 0xD1 )&& src[i+1]) {dest[j++] = src[++i] + 0x30;  }
    else dest[j++] = src[i];
  }
  dest[j] ='\0';
  return j;
}

/*--------------------------------------------------------------------------------------
  setup
  Called by the Arduino architecture before the main loop begins
--------------------------------------------------------------------------------------*/

void setup(void)
{
   
   // initialize DMD objects
   // scan DMD row each 700 us
    dmd.init(700);
    dmd.setBrightness(80);
   
}

**Reply by agganitk** [reply]
I changed to PA0
 #define DMD_PIN_SCLK PA0
SCLK pin is now toggling.

dont know why not toggling on selecting PA15 or PB4

**Reply by board707** [reply]
Is the DMD monochrome code now working?

**Reply by agganitk** [reply]
yes!!

**Reply by board707** [reply]
My congratulation!
Do you have any questions about displaying both fixed and scrolling text?

**Reply by agganitk** [reply]
I am yet to implement that.. Will come back..  I m sure I will have questions.. ;)
But overall library is amazing.. I see lot lot of work is done to make it.

**Reply by agganitk** [reply]
it is workingg!!!

**Reply by board707** [reply]
Thank you for your feedback
I added this effect to the examples, see the [dmd_rgb_dual_buf.ino](https://github.com/board707/DMD_STM32/blob/dev-V2/examples/STM32F4_RP2040/dmd_rgb_dual_buf/dmd_rgb_dual_buf.ino) example
